include/dirent.h

```diff
@@ -35,6 +44,24 @@ extern DIR *fdopendir(int fd) __DARWIN_ALIAS(fdopendir);
 extern DIR *fdopendir(int fd) __DARWIN_ALIAS_I(fdopendir);
 #endif
 
+extern DIR *__mpls_opendir(const char *name);
+extern int __mpls_closedir(DIR *dirp);
+extern int __mpls_dirfd(DIR *dirp);
+
+#define opendir(name)      __mpls_opendir(name)
```

Comment:
It is legal for applications to suppress macro expansion when calling `opendir()`, `readdir()` etc., since https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/dirent.h.html says

> The following shall be declared as functions and may also be defined as macros. Function prototypes shall be provided.

Rather than use `(opendir)(".")` for legacysupport to call the underlying implementation, I think it would be better to do this here:

```c
#define opendir __mpls_opendir
#define closedir __mpls_closedir

static inline struct dirent *__mpls_readdir(DIR *dirp) {
    return readdir(dirp->__mpls_dirp);
}
#define readdir __mpls_readdir

...
```

And then in `fdopendir.c` do

```
#undef DIR
#undef opendir
#undef closedir
#undef readdir
...

__MPLS_DIR *__mpls_fdopendir(int dirfd) {
    __MPLS_DIR *ret = malloc(sizeof(*ret);
    __pthread_fchdir(dirfd);
    ret->__mpls_dirp = opendir(".");
    ret->__mpls_dirfd = dirfd;
    __pthread_fchdir(-1);
    return ret;
}
```

Reply:
OK, but I don't understand in what way that is better. It's done.